rest/src/main/java/discord4j/rest/request/DefaultRouter.java

```diff
@@ -60,26 +68,66 @@ public DefaultRouter(RouterOptions routerOptions) {
     public DiscordWebResponse exchange(DiscordWebRequest request) {
         return new DiscordWebResponse(Mono.deferContextual(
                 ctx -> {
-                    RequestStream stream = getStream(request);
                     Sinks.One<ClientResponse> callback = Sinks.one();
-                    if (!stream.push(new RequestCorrelation<>(request, callback, ctx))) {
-                        callback.emitError(new DiscardedRequestException(request), FAIL_FAST);
-                    }
+                    housekeepIfNecessary();
+                    BucketKey bucketKey = BucketKey.of(request);
+                    streamMap.compute(bucketKey, (key, value) -> {
+                        RequestStream stream = value != null ? value : createStream(key, request);
+                        if (!stream.push(new RequestCorrelation<>(request, callback, ctx))) {
+                            callback.emitError(new DiscardedRequestException(request), FAIL_FAST);
+                        }
+                        return stream;
+                    });
                     return callback.asMono();
                 })
                 .checkpoint("Request to " + request.getDescription() + " [DefaultRouter]"), reactorResources);
     }
 
-    private RequestStream getStream(DiscordWebRequest request) {
-        return streamMap.computeIfAbsent(BucketKey.of(request),
-                k -> {
-                    if (log.isTraceEnabled()) {
-                        log.trace("Creating RequestStream with key {} for request: {} -> {}",
-                                k, request.getRoute().getUriTemplate(), request.getCompleteUri());
-                    }
-                    RequestStream stream = new RequestStream(k, routerOptions, httpClient, HEADER_STRATEGY);
-                    stream.start();
-                    return stream;
-                });
+    private RequestStream createStream(BucketKey bucketKey, DiscordWebRequest request) {
+        if (log.isTraceEnabled()) {
+            log.trace("Creating RequestStream with key {} for request: {} -> {}",
+                bucketKey, request.getRoute().getUriTemplate(), request.getCompleteUri());
+        }
+        RequestStream stream = new RequestStream(bucketKey, routerOptions, httpClient, HEADER_STRATEGY);
+        stream.start();
+        return stream;
+    }
+
+    private void housekeepIfNecessary() {
+        Instant lastTime = lastHousekeepingTime.get();
+        Instant now = Instant.now();
+        if (lastTime.plus(HOUSE_KEEPING_PERIOD).isAfter(now)) {
+            return;
+        }
+
+        tryHousekeep(now);
+    }
+
+    private void tryHousekeep(Instant now) {
+        boolean isTaken = someoneIsHousekeeping.getAndSet(true);
+        if (isTaken) {
+            return;
+        }
+
+        try {
+            doHousekeep(now);
+        } finally {
+            lastHousekeepingTime.set(Instant.now());
+            someoneIsHousekeeping.set(false);
+        }
+    }
```

Comment:
We could make this more compact by turning `lastHousekeepingTime` into a plain `Instant` field. What do you think?

```java
private void tryHousekeep(Instant now) {
    if (someoneIsHousekeeping.compareAndSet(false, true)) {
        try {
            doHousekeep(now);
        } finally {
            lastHousekeepingTime = Instant.now();
            someoneIsHousekeeping.set(false);
        }
    }
}
```